File: 01-Languages/Python/21-Solutions/solution-07.py
"""
Solution 07 - Mini Inventory System with SQLite
See: ../20-Exercises/README.md#exercise-07--mini-inventory-system-with-sqlite-advanced

Run with:
    python solution-07.py

Expected output:
    Items after adding three:
      {'id': 1, 'name': 'Widget', 'quantity': 10}
      {'id': 2, 'name': 'Gadget', 'quantity': 5}
      {'id': 3, 'name': 'Gizmo', 'quantity': 0}
    Updated Gadget quantity to 20
    Items after update:
      {'id': 1, 'name': 'Widget', 'quantity': 10}
      {'id': 2, 'name': 'Gadget', 'quantity': 20}
      {'id': 3, 'name': 'Gizmo', 'quantity': 0}
    Expected error caught: No item named 'Sprocket' exists
"""

import sqlite3


class ItemNotFoundError(Exception):
    pass


def init_db(conn: sqlite3.Connection) -> None:
    conn.execute(
        """
        CREATE TABLE items (
            id INTEGER PRIMARY KEY,
            name TEXT NOT NULL,
            quantity INTEGER NOT NULL DEFAULT 0
        )
        """
    )
    conn.commit()


def add_item(conn: sqlite3.Connection, name: str, quantity: int) -> None:
    # Using "?" placeholders (parameterized queries) instead of an f-string
    # means the driver treats name/quantity strictly as DATA, never as SQL
    # syntax - this is what prevents SQL injection, not just convenience.
    conn.execute(
        "INSERT INTO items (name, quantity) VALUES (?, ?)", (name, quantity)
    )
    conn.commit()


def update_quantity(conn: sqlite3.Connection, name: str, new_quantity: int) -> None:
    cursor = conn.execute(
        "UPDATE items SET quantity = ? WHERE name = ?", (new_quantity, name)
    )
    conn.commit()
    # cursor.rowcount tells us how many rows the UPDATE actually touched -
    # zero means the WHERE clause matched nothing, i.e. the name doesn't exist.
    if cursor.rowcount == 0:
        raise ItemNotFoundError(f"No item named '{name}' exists")


def list_items(conn: sqlite3.Connection) -> list[dict]:
    cursor = conn.execute("SELECT id, name, quantity FROM items ORDER BY id")
    # Column names come from cursor.description so this stays correct even
    # if the SELECT's column list changes later.
    columns = [col[0] for col in cursor.description]
    return [dict(zip(columns, row)) for row in cursor.fetchall()]


if __name__ == "__main__":
    connection = sqlite3.connect(":memory:")
    init_db(connection)

    add_item(connection, "Widget", 10)
    add_item(connection, "Gadget", 5)
    add_item(connection, "Gizmo", 0)

    print("Items after adding three:")
    for item in list_items(connection):
        print(f"  {item}")

    update_quantity(connection, "Gadget", 20)
    print("Updated Gadget quantity to 20")

    print("Items after update:")
    for item in list_items(connection):
        print(f"  {item}")

    try:
        update_quantity(connection, "Sprocket", 1)
    except ItemNotFoundError as e:
        print(f"Expected error caught: {e}")

    connection.close()

File: 01-Languages/Java/21-Solutions/Solution07.java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.ArrayList;
import java.util.List;

class ItemNotFoundException extends Exception {
    public ItemNotFoundException(String message) {
        super(message);
    }
}

record Item(int id, String name, int quantity) {}

public class Solution07 {

    static void initDb(Connection conn) throws SQLException {
        try (Statement stmt = conn.createStatement()) {
            stmt.execute("""
                CREATE TABLE items (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    name TEXT NOT NULL,
                    quantity INTEGER NOT NULL DEFAULT 0
                )
                """);
        }
    }

    static void addItem(Connection conn, String name, int quantity) throws SQLException {
        try (PreparedStatement stmt = conn.prepareStatement(
                "INSERT INTO items (name, quantity) VALUES (?, ?)")) {
            stmt.setString(1, name);
            stmt.setInt(2, quantity);
            stmt.executeUpdate();
        }
    }

    static void updateQuantity(Connection conn, String name, int newQuantity)
            throws SQLException, ItemNotFoundException {
        try (PreparedStatement stmt = conn.prepareStatement(
                "UPDATE items SET quantity = ? WHERE name = ?")) {
            stmt.setInt(1, newQuantity);
            stmt.setString(2, name);
            int rowsAffected = stmt.executeUpdate();
            if (rowsAffected == 0) {
                throw new ItemNotFoundException("No item named '" + name + "' exists");
            }
        }
    }

    static List<Item> listItems(Connection conn) throws SQLException {
        List<Item> items = new ArrayList<>();
        try (Statement stmt = conn.createStatement();
             ResultSet rs = stmt.executeQuery("SELECT id, name, quantity FROM items ORDER BY id")) {
            while (rs.next()) {
                items.add(new Item(rs.getInt("id"), rs.getString("name"), rs.getInt("quantity")));
            }
        }
        return items;
    }

    public static void main(String[] args) throws Exception {
        try (Connection conn = DriverManager.getConnection("jdbc:sqlite::memory:")) {
            initDb(conn);

            addItem(conn, "Widget", 10);
            addItem(conn, "Gadget", 5);
            addItem(conn, "Gizmo", 0);

            System.out.println("After adding 3 items:");
            for (Item item : listItems(conn)) {
                System.out.println("  " + item);
            }

            updateQuantity(conn, "Gizmo", 20);
            System.out.println("After updating Gizmo to 20:");
            for (Item item : listItems(conn)) {
                System.out.println("  " + item);
            }

            try {
                updateQuantity(conn, "Nonexistent", 5);
            } catch (ItemNotFoundException e) {
                System.out.println("Correctly caught: " + e.getMessage());
            }
        }
    }
}
